# Exploring Ebay Car Sales Data by Pavel Gladkevich
This project was completed as part of the Data Analyst series of [Dataquest](https://www.dataquest.io/directory/) on 06/18/19
<br/><br/>**Goal:** [Used Cars Database](https://www.kaggle.com/orgesleka/used-cars-database)

Over 370000 used cars scraped with Scrapy from Ebay-Kleinanzeigen. The content of the data is in german, so one has to translate it first if one can not speak german. Those fields are included: autos.csv:

- dateCrawled : when this ad was first crawled, all field-values are taken from this date
- name : "name" of the car
- seller : private or dealer
- offerType
- price : the price on the ad to sell the car
- abtest
- vehicleType
- yearOfRegistration : at which year the car was first registered
- gearbox 
- powerPS : power of the car in PS
- model
- kilometer : how many kilometers the car has driven
- monthOfRegistration : at which month the car was first registered
- fuelType
- brand
- notRepairedDamage : if the car has a damage which is not repaired yet
- dateCreated : the date for which the ad at ebay was created
- nrOfPictures : number of pictures in the ad (unfortunately this field contains everywhere a 0 and is thus useless (bug in crawler!) )
- postalCode
- lastSeenOnline : when the crawler saw this ad last online



In [1]:
import numpy as np
import pandas as pd

#load the downloaded csv file of used car data from file path
autos = pd.read_csv('/Users/pgladkevich/Desktop/coding/projects/datasets/used-cars-database/autos.csv', encoding = "Latin-1")
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
dateCrawled            371528 non-null object
name                   371528 non-null object
seller                 371528 non-null object
offerType              371528 non-null object
price                  371528 non-null int64
abtest                 371528 non-null object
vehicleType            333659 non-null object
yearOfRegistration     371528 non-null int64
gearbox                351319 non-null object
powerPS                371528 non-null int64
model                  351044 non-null object
kilometer              371528 non-null int64
monthOfRegistration    371528 non-null int64
fuelType               338142 non-null object
brand                  371528 non-null object
notRepairedDamage      299468 non-null object
dateCreated            371528 non-null object
nrOfPictures           371528 non-null int64
postalCode             371528 non-null int64
lastSeen              

In [2]:
#This will print some of the first and last rows in the pandas dataframe
autos

dateCrawled  \
0       2016-03-24 11:52:17   
1       2016-03-24 10:58:45   
2       2016-03-14 12:52:21   
3       2016-03-17 16:54:04   
4       2016-03-31 17:25:20   
5       2016-04-04 17:36:23   
6       2016-04-01 20:48:51   
7       2016-03-21 18:54:38   
8       2016-04-04 23:42:13   
9       2016-03-17 10:53:50   
10      2016-03-26 19:54:18   
11      2016-04-07 10:06:22   
12      2016-03-15 22:49:09   
13      2016-03-21 21:37:40   
14      2016-03-21 12:57:01   
15      2016-03-11 21:39:15   
16      2016-04-01 12:46:46   
17      2016-03-20 10:25:19   
18      2016-03-23 15:48:05   
19      2016-04-01 22:55:47   
20      2016-04-01 19:56:48   
21      2016-03-27 11:38:00   
22      2016-03-23 14:52:51   
23      2016-03-12 19:43:07   
24      2016-03-13 20:40:49   
25      2016-03-18 21:44:09   
26      2016-03-10 19:38:18   
27      2016-03-07 12:51:23   
28      2016-03-09 11:56:38   
29      2016-03-08 19:55:19   
...                     ...   
371498  2016-03-11 15:50:38   
371499  2016-03-08 13:52:00   
371500  2016-03-21 23:40:49   
371501  2016-03-26 18:36:54   
371502  2016-04-01 12:59:21   
371503  2016-04-01 01:55:58   
371504  2016-03-07 17:06:35   
371505  2016-04-03 19:56:57   
371506  2016-03-20 18:47:59   
371507  2016-03-16 17:06:19   
371508  2016-04-02 16:43:27   
371509  2016-03-11 23:40:32   
371510  2016-03-06 21:11:23   
371511  2016-03-31 18:50:37   
371512  2016-03-25 18:48:06   
371513  2016-03-21 15:36:22   
371514  2016-04-04 11:45:16   
371515  2016-03-09 11:36:55   
371516  2016-04-04 09:57:12   
371517  2016-03-28 13:48:07   
371518  2016-04-02 20:37:03   
371519  2016-03-09 13:37:43   
371520  2016-03-19 19:53:49   
371521  2016-03-27 20:36:20   
371522  2016-03-21 09:50:58   
371523  2016-03-14 17:48:27   
371524  2016-03-05 19:56:21   
371525  2016-03-19 18:57:12   
371526  2016-03-20 19:41:08   
371527  2016-03-07 19:39:19   

                                                     name  seller offerType  \
0                                              Golf_3_1.6  privat   Angebot   
1                                    A5_Sportback_2.7_Tdi  privat   Angebot   
2                          Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3                                      GOLF_4_1_4__3TÜRER  privat   Angebot   
4                          Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   
5       BMW_316i___e36_Limousine___Bastlerfahrzeug__Ex...  privat   Angebot   
6                             Peugeot_206_CC_110_Platinum  privat   Angebot   
7                            VW_Derby_Bj_80__Scheunenfund  privat   Angebot   
8                    Ford_C___Max_Titanium_1_0_L_EcoBoost  privat   Angebot   
9       VW_Golf_4_5_tuerig_zu_verkaufen_mit_Anhaengerk...  privat   Angebot   
10                                      Mazda_3_1.6_Sport  privat   Angebot   
11          Volkswagen_Passat_Variant_2.0_TDI_Comfortline  privat   Angebot   
12                      VW_Passat_Facelift_35i__"7Sitzer"  privat   Angebot   
13                         VW_PASSAT_1.9_TDI_131_PS_LEDER  privat   Angebot   
14      Nissan_Navara_2.5DPF_SE4x4_Klima_Sitzheizg_Blu...  privat   Angebot   
15                           KA_Lufthansa_Edition_450€_VB  privat   Angebot   
16                                            Polo_6n_1_4  privat   Angebot   
17                            Renault_Twingo_1.2_16V_Aut.  privat   Angebot   
18                       Ford_C_MAX_2.0_TDCi_DPF_Titanium  privat   Angebot   
19                      Mercedes_Benz_A_160_Classic_Klima  privat   Angebot   
20                      Volkswagen_Scirocco_1.4_TSI_Sport  privat   Angebot   
21      BMW_530i_TÜV_7/17_Scheckheftgepflegt_sehr_gute...  privat   Angebot   
22                          Opel_Meriva_1.Hand_TÜV_3.2018  privat   Angebot   
23                                           Stadtflitzer  privat   Angebot   
24                             MERCEDES_200E__TÜV_04/2016  privat   Angebot   
25                  BMW_530d_touring_Vo

In [3]:
print("hello")

hello


# Cleaning Column Names
As we can observe from the above data there are a total of 20 columns, with varied data types. The majority of the data is in string format, some have NaN values, and there are a couple of numerical columns. All of the columns are in camelCase, but the standard syntax for python is snake_case so we will start our analysis by converting the column names to this preferred case. Additionally some of the column names are unnecessarily verbose, so we will slightly change some of the wording to better fit the snake_case formatting.

In [4]:
# Using the dataframe.attribute to create column index object
autosColumns = autos.columns
print(autosColumns) 


Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


In [5]:
# Create copy of the index object
autos_columns = autosColumns.copy()
# Make four adjustments before we convert the rest to snake_case
autos_columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'kilometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'date_created', 'nmbr_pictures', 'postal_code',
       'last_seen']
autos.columns = autos_columns
autos.head(1)

date_crawled        name  seller offer_type  price abtest  \
0  2016-03-24 11:52:17  Golf_3_1.6  privat    Angebot    480   test   

  vehicle_type  registration_year  gearbox  power_ps model  kilometer  \
0          NaN               1993  manuell         0  golf     150000   

   registration_month fuel_type       brand unrepaired_damage  \
0                   0    benzin  volkswagen               NaN   

          date_created  nmbr_pictures  postal_code            last_seen  
0  2016-03-24 00:00:00              0        70435  2016-04-07 03:16:57

# Initial Exploration and Cleaning
From our above dataframe.head() we can tell that the columns of seller and offer_type have many singularly repeated values so it is possible that there is little variation of data types inside of these columns. 

In [6]:
autos.describe()

price  registration_year       power_ps      kilometer  \
count  3.715280e+05      371528.000000  371528.000000  371528.000000   
mean   1.729514e+04        2004.577997     115.549477  125618.688228   
std    3.587954e+06          92.866598     192.139578   40112.337051   
min    0.000000e+00        1000.000000       0.000000    5000.000000   
25%    1.150000e+03        1999.000000      70.000000  125000.000000   
50%    2.950000e+03        2003.000000     105.000000  150000.000000   
75%    7.200000e+03        2008.000000     150.000000  150000.000000   
max    2.147484e+09        9999.000000   20000.000000  150000.000000   

       registration_month  nmbr_pictures   postal_code  
count       371528.000000       371528.0  371528.00000  
mean             5.734445            0.0   50820.66764  
std              3.712412            0.0   25799.08247  
min              0.000000            0.0    1067.00000  
25%              3.000000            0.0   30459.00000  
50%              6.000000            0.0   49610.00000  
75%              9.000000            0.0   71546.00000  
max             12.000000            0.0   99998.00000

In [7]:
autos.describe(include="all")

date_crawled         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

        abtest vehicle_type  registration_year  gearbox       power_ps  \
count   371528       333659      371528.000000   351319  371528.000000   
unique       2            8                NaN        2            NaN   
top       test    limousine                NaN  manuell            NaN   
freq    192585        95894                NaN   274214            NaN   
mean       NaN          NaN        2004.577997      NaN     115.549477   
std        NaN          NaN          92.866598      NaN     192.139578   
min        NaN          NaN        1000.000000      NaN       0.000000   
25%        NaN          NaN        1999.000000      NaN      70.000000   
50%        NaN          NaN        2003.000000      NaN     105.000000   
75%        NaN          NaN        2008.000000      NaN     150.000000   
max        NaN          NaN        9999.000000      NaN   20000.000000   

         model      kilometer  registration_month fuel_type       brand  \
count   351044  371528.000000       371528.000000    338142      371528   
unique     251            NaN                 NaN         7          40   
top       golf            NaN                 NaN    benzin  volkswagen   
freq     30070            NaN                 NaN    223857       79640   
mean       NaN  125618.688228            5.734445       NaN         NaN   
std        NaN   40112.337051            3.712412       NaN         NaN   
min        NaN    5000.000000            0.000000       NaN         NaN   
25%        NaN  125000.000000            3.000000       NaN         NaN   
50%        NaN  150000.000000            6.000000       NaN         NaN   
75%        NaN  150000.000000            9.000000       NaN         NaN   
max        NaN  150000.000000           12.000000       NaN         NaN   

       unrepaired_damage         date_created  nmbr_pictures   postal_code  \
count             299468               371528       371528.0  371528.00000   
unique                 2                  114            NaN           NaN   
top                 nein  2016-04-03 00:00:00            NaN           NaN   
freq              263182                14450            NaN           NaN   
mean                 NaN                  NaN            0.0   50820.66764   
std                  NaN                  NaN            0.0   25799.08247   
min                  NaN                  NaN            0.0    1067.00000   
25%                  NaN                  NaN            0.0   30459.00000   
50%                  NaN                  NaN            0.0   49610.00000   
75%                  NaN                  NaN            0.0   71546.00000   
max                  NaN                  NaN            0.0   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-07 06:45:59  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                     NaN

The nmbr_pictures column has zeroes for every single numerical value, while the seller column has 371528 total values, but only 3 values that are not 'privat'. Likewise, the seller column has 371528 values but only 12 values are not equal to 'Angebot'. We will drop these columns from our dataframe

In [8]:
autos.drop(["seller","offer_type","nmbr_pictures"],axis=1)

date_crawled  \
0       2016-03-24 11:52:17   
1       2016-03-24 10:58:45   
2       2016-03-14 12:52:21   
3       2016-03-17 16:54:04   
4       2016-03-31 17:25:20   
5       2016-04-04 17:36:23   
6       2016-04-01 20:48:51   
7       2016-03-21 18:54:38   
8       2016-04-04 23:42:13   
9       2016-03-17 10:53:50   
10      2016-03-26 19:54:18   
11      2016-04-07 10:06:22   
12      2016-03-15 22:49:09   
13      2016-03-21 21:37:40   
14      2016-03-21 12:57:01   
15      2016-03-11 21:39:15   
16      2016-04-01 12:46:46   
17      2016-03-20 10:25:19   
18      2016-03-23 15:48:05   
19      2016-04-01 22:55:47   
20      2016-04-01 19:56:48   
21      2016-03-27 11:38:00   
22      2016-03-23 14:52:51   
23      2016-03-12 19:43:07   
24      2016-03-13 20:40:49   
25      2016-03-18 21:44:09   
26      2016-03-10 19:38:18   
27      2016-03-07 12:51:23   
28      2016-03-09 11:56:38   
29      2016-03-08 19:55:19   
...                     ...   
371498  2016-03-11 15:50:38   
371499  2016-03-08 13:52:00   
371500  2016-03-21 23:40:49   
371501  2016-03-26 18:36:54   
371502  2016-04-01 12:59:21   
371503  2016-04-01 01:55:58   
371504  2016-03-07 17:06:35   
371505  2016-04-03 19:56:57   
371506  2016-03-20 18:47:59   
371507  2016-03-16 17:06:19   
371508  2016-04-02 16:43:27   
371509  2016-03-11 23:40:32   
371510  2016-03-06 21:11:23   
371511  2016-03-31 18:50:37   
371512  2016-03-25 18:48:06   
371513  2016-03-21 15:36:22   
371514  2016-04-04 11:45:16   
371515  2016-03-09 11:36:55   
371516  2016-04-04 09:57:12   
371517  2016-03-28 13:48:07   
371518  2016-04-02 20:37:03   
371519  2016-03-09 13:37:43   
371520  2016-03-19 19:53:49   
371521  2016-03-27 20:36:20   
371522  2016-03-21 09:50:58   
371523  2016-03-14 17:48:27   
371524  2016-03-05 19:56:21   
371525  2016-03-19 18:57:12   
371526  2016-03-20 19:41:08   
371527  2016-03-07 19:39:19   

                                                     name  price   abtest  \
0                                              Golf_3_1.6    480     test   
1                                    A5_Sportback_2.7_Tdi  18300     test   
2                          Jeep_Grand_Cherokee_"Overland"   9800     test   
3                                      GOLF_4_1_4__3TÜRER   1500     test   
4                          Skoda_Fabia_1.4_TDI_PD_Classic   3600     test   
5       BMW_316i___e36_Limousine___Bastlerfahrzeug__Ex...    650     test   
6                             Peugeot_206_CC_110_Platinum   2200     test   
7                            VW_Derby_Bj_80__Scheunenfund      0     test   
8                    Ford_C___Max_Titanium_1_0_L_EcoBoost  14500  control   
9       VW_Golf_4_5_tuerig_zu_verkaufen_mit_Anhaengerk...    999     test   
10                                      Mazda_3_1.6_Sport   2000  control   
11          Volkswagen_Passat_Variant_2.0_TDI_Comfortline   2799  control   
12                      VW_Passat_Facelift_35i__"7Sitzer"    999  control   
13                         VW_PASSAT_1.9_TDI_131_PS_LEDER   2500  control   
14      Nissan_Navara_2.5DPF_SE4x4_Klima_Sitzheizg_Blu...  17999  control   
15                           KA_Lufthansa_Edition_450€_VB    450     test   
16                                            Polo_6n_1_4    300     test   
17                            Renault_Twingo_1.2_16V_Aut.   1750  control   
18                       Ford_C_MAX_2.0_TDCi_DPF_Titanium   7550     test   
19                      Mercedes_Benz_A_160_Classic_Klima   1850     test   
20                      Volkswagen_Scirocco_1.4_TSI_Sport  10400  control   
21      BMW_530i_TÜV_7/17_Scheckheftgepflegt_sehr_gute...   3699     test   
22                          Opel_Meriva_1.Hand_TÜV_3.2018   2900     test   
23                                           Stadtflitzer    450     test   
24                             MERCEDES_200E__TÜV_04/2016    500     test   
25                  BMW_530d_touring_Vollausstattung_NAVI   2500  control   
26           

In [9]:
autos.rename(columns={"kilometer" : "odometer_km"},inplace=True)
autos.head(1)

date_crawled        name  seller offer_type  price abtest  \
0  2016-03-24 11:52:17  Golf_3_1.6  privat    Angebot    480   test   

  vehicle_type  registration_year  gearbox  power_ps model  odometer_km  \
0          NaN               1993  manuell         0  golf       150000   

   registration_month fuel_type       brand unrepaired_damage  \
0                   0    benzin  volkswagen               NaN   

          date_created  nmbr_pictures  postal_code            last_seen  
0  2016-03-24 00:00:00              0        70435  2016-04-07 03:16:57

# Exploring Odometer and Price Columns

In [10]:
print("There are ",autos["odometer_km"].unique().shape[0]," different odometer readings")
print(autos["odometer_km"].describe())
print('\n')
print(autos["odometer_km"].value_counts())

There are  13  different odometer readings
count    371528.000000
mean     125618.688228
std       40112.337051
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64


150000    240797
125000     38067
100000     15920
90000      12523
80000      11053
70000       9773
60000       8669
50000       7615
5000        7069
40000       6376
30000       6041
20000       5676
10000       1949
Name: odometer_km, dtype: int64


It appears that there are rounded values so it is possible that sellers had to choose between preset options, or the data was altered before input. Additionally there are a many more vehicles that have 150,000 km on the odometer compared to any other field. This makes sense as this is a database of used car sales. This also makes it very likely that the 240,797 values are binned into 150,000 km and actually represent any odometer reading greater than this threshold. 

In [11]:
print("There are ",autos["price"].unique().shape[0]," different prices")
print(autos["price"].describe())
print('\n')
print(autos["price"].value_counts().head(20))

There are  5597  different prices
count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price, dtype: float64


0       10778
500      5670
1500     5394
1000     4649
1200     4594
2500     4438
600      3819
3500     3792
800      3784
2000     3432
999      3364
750      3203
650      3150
4500     3053
850      2946
2200     2936
700      2936
1800     2886
900      2874
950      2793
Name: price, dtype: int64


There are 10,778 zero values which seems like a lot at the first glance; however, there are 371,528 total values so this represents only 2.7% of the dataset so we can elect to remove these values since it is not needed for our analysis and could cause zero division errors or other problems down the line.  

In [12]:
print(autos["price"].value_counts().sort_index(ascending=False).head(20))
print(autos["price"].value_counts().sort_index().head(20))

2147483647     1
99999999      15
99000000       1
74185296       1
32545461       1
27322222       1
14000500       1
12345678       9
11111111      10
10010011       1
10000000       8
9999999        3
3895000        1
3890000        1
2995000        1
2795000        1
1600000        2
1300000        1
1250000        2
1234566        1
Name: price, dtype: int64
0     10778
1      1189
2        12
3         8
4         1
5        26
7         3
8         9
9         8
10       84
11        5
12        8
13        7
14        5
15       27
16        2
17        5
18        3
19        3
20       51
Name: price, dtype: int64


There is a natural progression of values from 0 to 20 dollars; however, on the other side of the spectrum after 389,500 there is a huge jump in monetary value that is skewing the statistics calculated for the averages of the dataset. These values are well above the rest and can definitely be qualitatively considered as outliers.

In [13]:
autos = autos[autos["price"].between(1,400000)]
autos.describe()

price  registration_year       power_ps    odometer_km  \
count  360641.000000      360641.000000  360641.000000  360641.000000   
mean     5904.931935        2004.433023     116.522697  125673.814125   
std      8998.294589          81.016330     189.874780   39820.293023   
min         1.000000        1000.000000       0.000000    5000.000000   
25%      1250.000000        1999.000000      72.000000  100000.000000   
50%      3000.000000        2004.000000     105.000000  150000.000000   
75%      7490.000000        2008.000000     150.000000  150000.000000   
max    399997.000000        9999.000000   20000.000000  150000.000000   

       registration_month  nmbr_pictures    postal_code  
count       360641.000000       360641.0  360641.000000  
mean             5.796041            0.0   50997.123938  
std              3.684421            0.0   25759.699797  
min              0.000000            0.0    1067.000000  
25%              3.000000            0.0   30823.000000  
50%              6.000000            0.0   49751.000000  
75%              9.000000            0.0   71672.000000  
max             12.000000            0.0   99998.000000

We can see what a great effect this had on our statistics calculated. It changed the the mean value to 5,904 dollars instead of 17,000 previously. The median of course stays the same.

# Exploring the Date Columns
The columns that contain date information are as follows:
- `date_crawled`: added by the crawler
- `last_seen`: added by the crawler
- `date_created`: from the website
- `registration_month`: from the website
- `registration_year`: from the website
<br><br>
The non registration dates are stored as strings


In [14]:
autos[['date_crawled','date_created','last_seen']][0:5]

date_crawled         date_created            last_seen
0  2016-03-24 11:52:17  2016-03-24 00:00:00  2016-04-07 03:16:57
1  2016-03-24 10:58:45  2016-03-24 00:00:00  2016-04-07 01:46:50
2  2016-03-14 12:52:21  2016-03-14 00:00:00  2016-04-05 12:47:46
3  2016-03-17 16:54:04  2016-03-17 00:00:00  2016-03-17 17:40:17
4  2016-03-31 17:25:20  2016-03-31 00:00:00  2016-04-06 10:17:21

In [15]:
# Sorting the column in order of earliest date to latest and by percentages of occurence
autos['date_crawled'].str[:10].value_counts(normalize=True,dropna=False).sort_index(ascending=True)

2016-03-05    0.025546
2016-03-06    0.014483
2016-03-07    0.035656
2016-03-08    0.033468
2016-03-09    0.034114
2016-03-10    0.032647
2016-03-11    0.032772
2016-03-12    0.036241
2016-03-13    0.015783
2016-03-14    0.036330
2016-03-15    0.033424
2016-03-16    0.030205
2016-03-17    0.031649
2016-03-18    0.013118
2016-03-19    0.035271
2016-03-20    0.036399
2016-03-21    0.035681
2016-03-22    0.032492
2016-03-23    0.032004
2016-03-24    0.029913
2016-03-25    0.032800
2016-03-26    0.031974
2016-03-27    0.030227
2016-03-28    0.035063
2016-03-29    0.034125
2016-03-30    0.033535
2016-03-31    0.031874
2016-04-01    0.034145
2016-04-02    0.035096
2016-04-03    0.038811
2016-04-04    0.037627
2016-04-05    0.012783
2016-04-06    0.003128
2016-04-07    0.001617
Name: date_crawled, dtype: float64

It appears that the data was crawled, in other words collected, from the website in a one month period from March to April 2016. The distribution of the data collection is roughly uniform with almost every date having an approximately equal percentage of the total number of posts.

In [16]:
autos['date_created'].str[:10].value_counts(normalize=True,dropna=False).sort_index(ascending=True)

2014-03-10    0.000003
2015-03-20    0.000003
2015-06-11    0.000003
2015-06-18    0.000003
2015-08-07    0.000003
2015-08-10    0.000003
2015-09-04    0.000006
2015-09-09    0.000003
2015-10-14    0.000003
2015-11-02    0.000003
2015-11-08    0.000003
2015-11-10    0.000003
2015-11-12    0.000003
2015-11-13    0.000003
2015-11-17    0.000003
2015-11-23    0.000006
2015-11-24    0.000006
2015-12-05    0.000008
2015-12-06    0.000003
2015-12-17    0.000003
2015-12-27    0.000003
2015-12-28    0.000003
2015-12-30    0.000008
2016-01-02    0.000011
2016-01-03    0.000003
2016-01-06    0.000003
2016-01-07    0.000008
2016-01-08    0.000003
2016-01-10    0.000011
2016-01-13    0.000008
                ...   
2016-03-09    0.034161
2016-03-10    0.032523
2016-03-11    0.032805
2016-03-12    0.036077
2016-03-13    0.017056
2016-03-14    0.035265
2016-03-15    0.033349
2016-03-16    0.030426
2016-03-17    0.031247
2016-03-18    0.014022
2016-03-19    0.034303
2016-03-20    0.036488
2016-03-21 

In [17]:
autos['last_seen'].str[:10].value_counts(normalize=True,dropna=False).sort_index(ascending=True)

2016-03-05    0.001264
2016-03-06    0.004098
2016-03-07    0.005202
2016-03-08    0.007939
2016-03-09    0.009824
2016-03-10    0.011460
2016-03-11    0.012955
2016-03-12    0.023239
2016-03-13    0.008410
2016-03-14    0.012176
2016-03-15    0.016324
2016-03-16    0.016418
2016-03-17    0.028702
2016-03-18    0.006888
2016-03-19    0.016329
2016-03-20    0.019884
2016-03-21    0.020025
2016-03-22    0.020508
2016-03-23    0.018015
2016-03-24    0.019163
2016-03-25    0.019000
2016-03-26    0.015958
2016-03-27    0.016720
2016-03-28    0.022188
2016-03-29    0.023284
2016-03-30    0.023724
2016-03-31    0.024243
2016-04-01    0.023896
2016-04-02    0.024967
2016-04-03    0.025310
2016-04-04    0.025535
2016-04-05    0.126968
2016-04-06    0.218949
2016-04-07    0.130434
Name: last_seen, dtype: float64

A record of when the crawler last saw the listing, which is likely time of sale for each vehicle. It is also possible that the owner of the advertisement decided not to sell the vehicle and simply took down the add, but this is unlikely to be a major percentage of the listings.

In [18]:
autos["registration_year"].describe()

count    360641.000000
mean       2004.433023
std          81.016330
min        1000.000000
25%        1999.000000
50%        2004.000000
75%        2008.000000
max        9999.000000
Name: registration_year, dtype: float64

It looks like there are some inconsistencies in the registration data, as it is impossible for a car to have been registered in the year of 1000 or 9999.

# Dealing With Incorrect Registration Data
It is unrealistic for a car to have been registered before the start of the 19th century or after the last created advertisement in 2016. Thus we will exclude any data that falls outside of this 116 year interval.

In [19]:
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].value_counts(normalize=True).sort_index(ascending = True)

1910    0.000167
1911    0.000003
1923    0.000009
1925    0.000003
1927    0.000006
1928    0.000006
1929    0.000023
1930    0.000012
1931    0.000012
1932    0.000012
1933    0.000014
1934    0.000014
1935    0.000014
1936    0.000020
1937    0.000035
1938    0.000023
1939    0.000012
1940    0.000006
1941    0.000012
1942    0.000006
1943    0.000014
1944    0.000006
1945    0.000014
1946    0.000006
1947    0.000014
1948    0.000009
1949    0.000009
1950    0.000061
1951    0.000049
1952    0.000032
          ...   
1987    0.001895
1988    0.002671
1989    0.003632
1990    0.007220
1991    0.007664
1992    0.008354
1993    0.009513
1994    0.013402
1995    0.025959
1996    0.030029
1997    0.040468
1998    0.049690
1999    0.063551
2000    0.066698
2001    0.056957
2002    0.054289
2003    0.056556
2004    0.056172
2005    0.062668
2006    0.057707
2007    0.050498
2008    0.046255
2009    0.044706
2010    0.035394
2011    0.034604
2012    0.026997
2013    0.017657
2014    0.0137

In [20]:
print("The percentage of data remaining from the original autos database is", autos.shape[0]/360641)

The percentage of data remaining from the original autos database is 0.9612495528794563


As expected, there are far more cars registered in more recent years than those prior to 1990. The distribution appears to be roughly gaussian with a mode at 2000; however, further statistial methods would need to be applied to determine how well the data fits this distribution. Lastly, it appears that less than 4% of the data was deleted as a result of removing the inconsistencies.

# Exploring Price by Brand

We will take a look at what brands are in the database and which are the most prevalent in the used car database.

In [21]:
autos['brand'].value_counts(normalize=True,dropna=False)

volkswagen        0.211697
bmw               0.109872
opel              0.106408
mercedes_benz     0.096842
audi              0.089542
ford              0.068916
renault           0.047515
peugeot           0.030153
fiat              0.025690
seat              0.018661
skoda             0.015687
mazda             0.015384
smart             0.014331
citroen           0.013950
nissan            0.013598
toyota            0.012932
hyundai           0.009972
sonstige_autos    0.009499
mini              0.009384
volvo             0.009147
mitsubishi        0.008236
honda             0.007532
kia               0.006914
suzuki            0.006363
alfa_romeo        0.006309
porsche           0.006216
chevrolet         0.005022
chrysler          0.003863
dacia             0.002495
jeep              0.002192
land_rover        0.002166
daihatsu          0.002161
subaru            0.002117
jaguar            0.001734
saab              0.001465
daewoo            0.001457
trabant           0.001408
l

The german manufacturers alone acount for the top 5 brands in the used car market and combined are nearly sixty percent of the advertisements. This is logical as we are using the german ebay as our data source. We will limit our analysis exclusively to these brands as they are the majority of the market and the most likely to appear in the future as well, thus improving the potential conclusions drawn. 

In [34]:
# Create empty dictionary that will contain the brand and its mean price
brand_mean_price = {}
brand_names = autos["brand"].value_counts(normalize=True,dropna=False).index[0:5]

for brand in brand_names:
    brand_mean_price[brand] = autos[autos["brand"] == brand].price.mean().round(2)
    
print(brand_mean_price)
    

{'volkswagen': 5400.19, 'bmw': 8459.14, 'opel': 2971.9, 'mercedes_benz': 8562.36, 'audi': 9086.28}


It appears that Opel has the lowest mean resale value at just under 3,000 dollars while Audi has the highest at a little over 9,000 dollars. BMW and Mercedes are fairly equal to Audi, while Volkswaggen is in the middle. The results follow what we would expect as the clustering of mean price is by budget, standard, and luxury car brands.

# Storing Aggregate Data in a Dataframe

We will calculate the mean mileage for the five brands that we analysed previously, as well as create a dataframe holding this information combined with the price means.

In [50]:
brand_mean_mileage = {}

# Create a separate dictionary with the mileage
for brand in brand_names:
    brand_mean_mileage[brand] = autos[autos["brand"] == brand].odometer_km.mean().round(2)

# Make two separate series out of the dictionaries
price_series = pd.Series(brand_mean_price)
mileage_series = pd.Series(brand_mean_mileage)
df = pd.DataFrame(price_series, columns=["Mean Price"])
df

Mean Price
volkswagen        5400.19
bmw               8459.14
opel              2971.90
mercedes_benz     8562.36
audi              9086.28

In [53]:
# Add the mileage data
df["Mileage"] = mileage_series
df

Mean Price    Mileage
volkswagen        5400.19  128386.04
bmw               8459.14  132796.74
opel              2971.90  128722.62
mercedes_benz     8562.36  130569.97
audi              9086.28  129443.96

# Conclusion

The mileage is all roughly equal and the variations within mileage do not appear to be driving the differences in price. Instead the brand's name itself is more highly correlated with the mean price. Lastly there appears to be a slight trend for the more expensive brands to have higher mileage, perhaps as these cars are more desireable even at a higher mileage to the average consumer. 